File: packages/i18nify-js/src/modules/dateTime/__tests__/getRelativeTime.test.ts
import { getRelativeTime } from '../index';

describe('dateTime - getRelativeTime', () => {
  const now = new Date();

  test('returns correct relative time for seconds', () => {
    const thirtySecondsAgo = new Date(now.getTime() - 30 * 1000);
    expect(getRelativeTime(thirtySecondsAgo)).toBe('30 seconds ago');
    const inThirtySeconds = new Date(now.getTime() + 30 * 1000);
    expect(getRelativeTime(inThirtySeconds)).toBe('in 30 seconds');
  });

  test('returns correct relative time for minutes', () => {
    const fiveMinutesAgo = new Date(now.getTime() - 5 * 60 * 1000);
    expect(getRelativeTime(fiveMinutesAgo)).toBe('5 minutes ago');
    const inFiveMinutes = new Date(now.getTime() + 5 * 60 * 1000);
    expect(getRelativeTime(inFiveMinutes)).toBe('in 5 minutes');
  });

  test('returns correct relative time for hours', () => {
    const twoHoursAgo = new Date(now.getTime() - 2 * 60 * 60 * 1000);
    expect(getRelativeTime(twoHoursAgo)).toBe('2 hours ago');
    const inTwoHours = new Date(now.getTime() + 2 * 60 * 60 * 1000);
    expect(getRelativeTime(inTwoHours)).toBe('in 2 hours');
  });

  test('returns correct relative time for days', () => {
    const threeDaysAgo = new Date(now.getTime() - 3 * 24 * 60 * 60 * 1000);
    expect(getRelativeTime(threeDaysAgo)).toBe('3 days ago');
    const inThreeDays = new Date(now.getTime() + 3 * 24 * 60 * 60 * 1000);
    expect(getRelativeTime(inThreeDays)).toBe('in 3 days');
  });

  test('returns correct relative time for days for undefined options', () => {
    const threeDaysAgo = new Date(now.getTime() - 3 * 24 * 60 * 60 * 1000);
    expect(getRelativeTime(threeDaysAgo, undefined)).toBe('3 days ago');
  });

  test('handles different locales', () => {
    const oneDayAgo = new Date(now.getTime() - 24 * 60 * 60 * 1000);
    expect(getRelativeTime(oneDayAgo, { locale: 'en-US' })).toBe('1 day ago');
    expect(getRelativeTime(oneDayAgo, { locale: 'fr-FR' })).toBe(
      'il y a 1 jour',
    );
  });

  test('throws an error for invalid date inputs', () => {
    expect(() => getRelativeTime('invalid-date')).toThrow(
      'Date format not recognized',
    );
  });

  test('throws a generic error message for invalid intl options', () => {
    const oneDayAgo = new Date(now.getTime() - 24 * 60 * 60 * 1000);
    expect(() =>
      getRelativeTime(new Date(), {
        locale: 'en-US',
        intlOptions: { style: 'dummy' } as any,
        baseDate: oneDayAgo,
      }),
    ).toThrow(
      'Error: An error occurred while creating the RelativeTimeFormat instance: Value dummy out of range for Intl.RelativeTimeFormat options property style. Please ensure the provided options are valid and try again.',
    );
  });

  test('handles non-Error instance in catch block', () => {
    const oneDayAgo = new Date(now.getTime() - 24 * 60 * 60 * 1000);
    // Create a custom error that's not an instance of Error
    const customError = { message: 'custom error' };

    // Mock the Intl.RelativeTimeFormat constructor to throw our custom error
    const originalRTF = Intl.RelativeTimeFormat;
    (Intl as any).RelativeTimeFormat = class {
      constructor() {
        throw customError;
      }
    };

    expect(() =>
      getRelativeTime(new Date(), {
        locale: 'en-US',
        baseDate: oneDayAgo,
      }),
    ).toThrow('An unknown error occurred. Error details: [object Object]');

    // Restore original implementation
    (Intl as any).RelativeTimeFormat = originalRTF;
  });

  test('returns correct relative time for weeks', () => {
    const twoWeeksAgo = new Date(now.getTime() - 2 * 7 * 24 * 60 * 60 * 1000);
    expect(getRelativeTime(twoWeeksAgo)).toBe('2 weeks ago');
    const inTwoWeeks = new Date(now.getTime() + 2 * 7 * 24 * 60 * 60 * 1000);
    expect(getRelativeTime(inTwoWeeks)).toBe('in 2 weeks');
  });

  test('returns correct relative time for months', () => {
    const threeMonthsAgo = new Date(
      now.getTime() - 3 * 30 * 24 * 60 * 60 * 1000,
    );
    expect(getRelativeTime(threeMonthsAgo)).toBe('3 months ago');
    const inThreeMonths = new Date(
      now.getTime() + 3 * 30 * 24 * 60 * 60 * 1000,
    );
    expect(getRelativeTime(inThreeMonths)).toBe('in 3 months');
  });

  test('returns correct relative time for years', () => {
    const oneYearAgo = new Date(now.getTime() - 365 * 24 * 60 * 60 * 1000);
    expect(getRelativeTime(oneYearAgo)).toBe('1 year ago');
    const inOneYear = new Date(now.getTime() + 365 * 24 * 60 * 60 * 1000);
    expect(getRelativeTime(inOneYear)).toBe('in 12 months');
  });
});

File: packages/i18nify-go/modules/date_time/get_relative_time_test.go
package date_time

import (
	"testing"
	"time"
)

func TestGetRelativeTime(t *testing.T) {
	now := time.Now()

	tests := []struct {
		name        string
		setup       func() (string, string) // returns date and baseDate
		options     *DateTimeOptions
		expected    string
		expectError bool
	}{
		// Seconds tests
		{
			name: "returns correct relative time for seconds - past",
			setup: func() (string, string) {
				date := now.Add(-30 * time.Second)
				return date.Format(time.RFC3339), now.Format(time.RFC3339)
			},
			expected: "30 seconds ago",
		},
		{
			name: "returns correct relative time for seconds - future",
			setup: func() (string, string) {
				date := now.Add(30 * time.Second)
				return date.Format(time.RFC3339), now.Format(time.RFC3339)
			},
			expected: "in 30 seconds",
		},

		// Minutes tests
		{
			name: "returns correct relative time for minutes - past",
			setup: func() (string, string) {
				date := now.Add(-5 * time.Minute)
				return date.Format(time.RFC3339), now.Format(time.RFC3339)
			},
			expected: "5 minutes ago",
		},
		{
			name: "returns correct relative time for minutes - future",
			setup: func() (string, string) {
				date := now.Add(5 * time.Minute)
				return date.Format(time.RFC3339), now.Format(time.RFC3339)
			},
			expected: "in 5 minutes",
		},

		// Hours tests
		{
			name: "returns correct relative time for hours - past",
			setup: func() (string, string) {
				date := now.Add(-2 * time.Hour)
				return date.Format(time.RFC3339), now.Format(time.RFC3339)
			},
			expected: "2 hours ago",
		},
		{
			name: "returns correct relative time for hours - future",
			setup: func() (string, string) {
				date := now.Add(2 * time.Hour)
				return date.Format(time.RFC3339), now.Format(time.RFC3339)
			},
			expected: "in 2 hours",
		},

		// Days tests
		{
			name: "returns correct relative time for days - past",
			setup: func() (string, string) {
				date := now.Add(-3 * 24 * time.Hour)
				return date.Format(time.RFC3339), now.Format(time.RFC3339)
			},
			expected: "3 days ago",
		},
		{
			name: "returns correct relative time for days - future",
			setup: func() (string, string) {
				date := now.Add(3 * 24 * time.Hour)
				return date.Format(time.RFC3339), now.Format(time.RFC3339)
			},
			expected: "in 3 days",
		},

		// Weeks tests
		{
			name: "returns correct relative time for weeks - past",
			setup: func() (string, string) {
				date := now.Add(-2 * 7 * 24 * time.Hour)
				return date.Format(time.RFC3339), now.Format(time.RFC3339)
			},
			expected: "2 weeks ago",
		},
		{
			name: "returns correct relative time for weeks - future",
			setup: func() (string, string) {
				date := now.Add(2 * 7 * 24 * time.Hour)
				return date.Format(time.RFC3339), now.Format(time.RFC3339)
			},
			expected: "in 2 weeks",
		},

		// Months tests
		{
			name: "returns correct relative time for months - past",
			setup: func() (string, string) {
				date := now.Add(-3 * 30 * 24 * time.Hour)
				return date.Format(time.RFC3339), now.Format(time.RFC3339)
			},
			expected: "3 months ago",
		},
		{
			name: "returns correct relative time for months - future",
			setup: func() (string, string) {
				date := now.Add(3 * 30 * 24 * time.Hour)
				return date.Format(time.RFC3339), now.Format(time.RFC3339)
			},
			expected: "in 3 months",
		},

		// Years tests
		{
			name: "returns correct relative time for years - past",
			setup: func() (string, string) {
				date := now.Add(-365 * 24 * time.Hour)
				return date.Format(time.RFC3339), now.Format(time.RFC3339)
			},
			expected: "1 Year ago",
		},
		{
			name: "returns correct relative time for years - future",
			setup: func() (string, string) {
				date := now.Add(365 * 24 * time.Hour)
				return date.Format(time.RFC3339), now.Format(time.RFC3339)
			},
			expected: "in 1 Year",
		},

		// Error cases
		{
			name: "throws error for invalid date input",
			setup: func() (string, string) {
				return "invalid-date", now.Format(time.RFC3339)
			},
			expectError: true,
		},

		// Locale tests
		{
			name: "handles different locales - en-US",
			setup: func() (string, string) {
				date := now.Add(-24 * time.Hour)
				return date.Format(time.RFC3339), now.Format(time.RFC3339)
			},
			expected: "1 Day ago",
		},
		// locale to be supported later
		//{
		//	name: "handles different locales - fr-FR",
		//	setup: func() (string, string) {
		//		date := now.Add(-24 * time.Hour)
		//		return date.Format(time.RFC3339), now.Format(time.RFC3339)
		//	},
		//	options: &DateTimeOptions{
		//		Locale: "fr-FR",
		//	},
		//	expected: "il y a 1 jour",
		//},

		// Undefined options test
		{
			name: "returns correct relative time for days with undefined options",
			setup: func() (string, string) {
				date := now.Add(-3 * 24 * time.Hour)
				return date.Format(time.RFC3339), now.Format(time.RFC3339)
			},
			options:  nil,
			expected: "3 days ago",
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			date, baseDate := tt.setup()
			options := DateTimeOptions{
				BaseDate: baseDate,
			}

			result, err := GetRelativeTime(date, options)

			if tt.expectError {
				if err == nil {
					t.Error("expected error but got none")
				}
				return
			}

			if err != nil {
				t.Errorf("unexpected error: %v", err)
				return
			}

			if result != tt.expected {
				t.Errorf("got %q, want %q", result, tt.expected)
			}
		})
	}
}
